### PR TITLE
Rewrite documentation

### LINKED ISSUE #63: Update Documentation
- [x] Adjust credit section with new contributions
- [x] Document the use of `group` and `name` columns in params
- [x] Improve the order and overall structure
- [x] More description of constraints, including tutorials and corner cases. 
- [x] Write section about contributing
    - [x] Describe the PR workflow
    - [x] Explain how to run tests and activate pre-commit hooks locally
    - [x] Explain importance of using `environment.yml`
    - [x] Explain how important it is that functions have no side effects.

### PR DESCRIPTION
**Pull request by janosg**
## Current Situation

closes #63 

A lot of documentation is written on a too abstract level for most users.

## Changes

- [x] Add getting started section with ordered logit notebook
- [x] Add separate sections for each of the more complex arguments of maximize and minimize
    - [x] params
    - [x] constraints (a notebook with many examples)
    - [x] algo options and algorithms
    - [x] dashboard and dashboard options
    - [x] general options
- [x] Improve the overall structure
- [x] Update documentation where outdated
- [x] Fix typos and introduce new ones
- [x] Write guides for contributions
    - [x] Explain editable installs

- [x] Add faq section
    - [x] positive definiteness problems
- [x] Document the utilities module
- [x] Add the constraint killers

### CODE REVIEW

File: docs/source/getting_started/index.rst [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+Getting Started
+===============
+
+.. toctree::
+   :maxdepth: 1
+
+
+   introduction
+   installation
+   ordered_logit_example.ipynb
```
Comment (roecla) [inline docs/source/getting_started/index.rst]: the ordered logit example seems to not have been committed yet. Possibly, because of .gitignore?

File: docs/source/optimization/index.rst [hunk 2]
```diff
@@ -0,0 +1,13 @@
+============
```
Comment (roecla) [inline docs/source/optimization/index.rst]: inconsistent capitalization.

File: docs/source/optimization/interface.rst [hunk 3]
```diff
@@ -0,0 +1,23 @@
+minimize and maximize
+=====================
+
+.. currentmodule:: estimagic.optimization.optimize
+
+Both optimization functions have the exactly same interface. In fact, ``maximize``
+ just switches the sign of your criterion function and calls ``minimize``.
+
+
+.. autofunction:: minimize
```
Comment (roecla) [inline docs/source/optimization/interface.rst]: The constraints argument is not properly rendered. It just says "See for details" without a link to click on.

File: docs/source/optimization/algorithms.rst [hunk 4]
```diff
@@ -0,0 +1,78 @@
+.. _list_of_algorithms:
+
+algo_options
+============
+
+
+``algo_options`` is a dictionary with optional keyword arguments that are passed
+to the optimizer. This includes tolerances for the termination criteria,
+parameters that determine how greedy the optimizer how large the stepsize for
```
Comment (roecla) [inline docs/source/optimization/algorithms.rst]: Maybe explain what you mean with "how greedy the optimizer". Also there is av verb missing I think.

File: docs/source/optimization/algorithms.rst [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+have a very difficult optimization problem. If so, you can find all available
+options at the following links (depending on the origin of the algorithm.)
+
+`pygmo <https://tinyurl.com/y3bgsl4z>`__.
+`nlopt <https://tinyurl.com/y3dsmaz7>`__.
+`scipy <https://tinyurl.com/y54nmedo>`__.
+
+
+``estimagic`` supports the following algorithms for optimization. To learn more
+about the corresponding ``algo_options``, follow the accompanying links.
```
Comment (roecla) [inline docs/source/optimization/algorithms.rst]: The links are still missing.

File: docs/source/optimization/params.rst [hunk 6]
```diff
@@ -0,0 +1,91 @@
+.. _params:
+
+====================
+The params DataFrame
+====================
+
+The ``params`` DataFrame is an important concept in estimagic. It collects
+information on the dimensionality of the optimization problem, lower and upper
+bounds, categories of parameters and valid ranges for randomly sampled parameter
+vectors. Moreover, it is the mandatory first argument of any criterion function
```
Comment (roecla) [inline docs/source/optimization/params.rst]: the criterion function only use Series ... at least that's what's written in the documentation of minimize and maximize.

File: docs/source/optimization/params.rst [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+optimized with estimagic.
+
+
+If you haven't done so yet, you should check out our example notebook for an
+ordered logit model, so you see one small params DataFrame in action.
```
Comment (roecla) [inline docs/source/optimization/params.rst]: add a reference so the reader can get to the example with one click

File: docs/source/differentiation/background.rst [hunk 8]
```diff
@@ -0,0 +1,77 @@
+Background and Methods
+======================
+
+In this section we will explain the mathematical background of forward, backward and
+central differences. The main ideas in this chapter are taken from this [Dennis]_ and
+Schneidler. Our toolkit is developed for multivariate functions but can of
```
Comment (roecla) [inline docs/source/differentiation/background.rst]: reference to the Schneidler book (?) is missing.

File: docs/source/differentiation/differentiation.rst [hunk 9]
```diff
@@ -0,0 +1,49 @@
+
+Gradient, Jacobian and Hessian
+==============================
+
+Estimagic wraps `numdifftools <https://pypi.org/project/numdifftools/>`_ to provide functions to calculate very precise gradients, jacobians and hessians of functions. The precision is achieved by evaluating numerical derivatives at different step sizes and using richardson extrapolations. While this increases the computational cost, it works for any function, whereas other approaches that would yield a similar precision, like complex step derivatives have stronger requirements.
+
+
+.. automodule:: estimagic.differentiation.differentiation
+    :members:
```
Comment (roecla) [inline docs/source/differentiation/differentiation.rst]: the Returns part of this docstring is not very informative ("the index is the index of params_sr")

File: docs/source/inference/moments_inference.rst [hunk 10]
```diff
@@ -1,20 +1,10 @@
...
+========================
 
 .. currentmodule:: estimagic.inference.moment_covs
 
-In this section we document how calculate standard errors of a (GMM) estimator. We use
-the notation and formulations provided in section 13.25 of
+In this section we document how calculate standard errors of a GMM or MSM estimator.
+We use the notation and formulations provided in section 13.25 of
 
 Bruce E. Hansen - Econometrics (https://www.ssc.wisc.edu/~bhansen/econometrics)
```
Comment (roecla) [inline docs/source/inference/moments_inference.rst]: different way to write the reference from how it is done in numerical differentiation. We should choose one for consistency but I don't have a preference.

File: docs/source/inference/likelihood_inference.rst [hunk 11]
```diff
@@ -0,0 +1,6 @@
+Maximum Likelihood
```
Comment (roecla) [inline docs/source/inference/likelihood_inference.rst]: This section is also inconsistent with the way that resources are cited.

**Comment by roecla** [comment]
A more general comment: Maybe we want to rename the visualization module to something like "Post-Estimation Tools". 
Firstly, it could confuse users because the dashboard is also a visualization. Secondly, we might have other tools in the future that wouldn't be considered visualizations (LaTeX tables being the first coming to my mind).

File: docs/source/getting_started/index.rst (hunk 1 above)
Comment (janosg) [reply docs/source/getting_started/index.rst]: Done.

File: docs/source/getting_started/introduction.rst [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+Inference
+=========
+
+Estimagic provides functions to calculate standard errors for Maximum
+Likelihood, GMM, Method of Simulated Moments and Indirect inference estimators.
+
+Visualization
+=============
+
+Interactive plots that help you to compare large numbers of estimated
```
[the remark this answers is not in the document]
Comment (janosg) [reply docs/source/getting_started/introduction.rst]: Done.

File: docs/source/optimization/index.rst (hunk 2 above)
Comment (janosg) [reply docs/source/optimization/index.rst]: Done.

File: docs/source/optimization/algorithms.rst (hunk 4 above)
Comment (janosg) [reply docs/source/optimization/algorithms.rst]: I don't want to add more explanations. This is just one example and there are two more, so people who don't understand it can continue.

File: docs/source/optimization/algorithms.rst (hunk 5 above)
Comment (janosg) [reply docs/source/optimization/algorithms.rst]: Done.

File: docs/source/optimization/params.rst (hunk 6 above)
Comment (janosg) [reply docs/source/optimization/params.rst]: This was changed months ago!!! I found one outdated docstring, but hope that's it.

File: docs/source/optimization/params.rst (hunk 7 above)
Comment (janosg) [reply docs/source/optimization/params.rst]: Had to find out how that works, but now it's done. Also in other places of the documentation.

File: docs/source/differentiation/background.rst (hunk 8 above)
Comment (janosg) [reply docs/source/differentiation/background.rst]: That took time. It's just one book, but not by Dennis and Schneidler, but Dennis and Schnabel!

File: docs/source/inference/moments_inference.rst (hunk 10 above)
Comment (janosg) [reply docs/source/inference/moments_inference.rst]: Now citations are done properly using sphinxcontrib.bibtex and a `refs.bib` file that collects all references.

File: docs/source/inference/likelihood_inference.rst (hunk 11 above)
Comment (janosg) [reply docs/source/inference/likelihood_inference.rst]: Also properly cited now.

File: docs/source/differentiation/differentiation.rst (hunk 9 above)
Comment (janosg) [reply docs/source/differentiation/differentiation.rst]: fixed.

File: docs/source/optimization/interface.rst (hunk 3 above)
Comment (janosg) [reply docs/source/optimization/interface.rst]: Should be fixed.

**Comment by janosg** [comment]
I think we rename it to post-estimation tools when there is more than just a few plotting functions. Currently I don't want to add new features but make sure everything we have is working really well.
